backpack/core/derivatives/adaptive_avg_pool_nd.py

```diff
@@ -0,0 +1,112 @@
+"""Implements the derivatives for AdaptiveAvgPool."""
+from typing import List, Tuple, Union
+
+from torch import Size
+from torch.nn import AdaptiveAvgPool1d, AdaptiveAvgPool2d, AdaptiveAvgPool3d
+
+from backpack.core.derivatives.avgpoolnd import AvgPoolNDDerivatives
+
+
+class AdaptiveAvgPoolNDDerivatives(AvgPoolNDDerivatives):
+    """Implements the derivatives for AdaptiveAvgPool."""
+
+    def __init__(self, N: int):
+        """Initialization.
+
+        Args:
+            N: number of dimensions
+        """
+        super(AdaptiveAvgPoolNDDerivatives, self).__init__(N=N)
+
+    def check_parameters(
+        self, module: Union[AdaptiveAvgPool1d, AdaptiveAvgPool2d, AdaptiveAvgPool3d]
+    ) -> None:
+        """Checks if the parameters are supported.
+
+        Specifically checks if input shape is multiple of output shape.
+        In this case, there are parameters for AvgPoolND that are equivalent.
+
+        https://stackoverflow.com/questions/53841509/how-does-adaptive-pooling-in-pytorch-work/63603993#63603993 # noqa: B950
+
+        Args:
+            module: module to check
+
+        Raises:
+            NotImplementedError: if the given shapes do not match
+        """
+        shape_input: Size = module.input0.shape
+        shape_target: Size = module.output.shape
+
+        # check length of input shape
+        if not len(shape_input) == (self.N + 2):
+            raise NotImplementedError(
+                f"input must be (batch_size, C, ...) with ... {self.N} dimensions"
+            )
+        # check length of target shape
+        if not len(shape_target) == (self.N + 2):
+            raise NotImplementedError(
+                f"shape of target should have {self.N} dimensions"
+            )
```

Comment:
Is this check necessary?

Reply:
Do you mean that violating inputs would be caught by the forward pass of `AvgPoolNd`?

Reply:
Exactly. And consequently, outputs should be fine as well.

Reply:
I think it is sufficient to only check the input shape and assume the forward pass preserves the dimensions.

I vote for keeping the input check, as adaptive average pooling also works on unbatched inputs:
```python
from torch import rand
from torch.nn.functional import adaptive_avg_pool2d

# unbatched image
input = rand(3, 20, 20)
output = adaptive_avg_pool2d(input, output_size=(3, 3))
print(output.shape) # -> torch.Size([3, 3, 3])

# batched image
input = rand(1, 3, 20, 20)
output = adaptive_avg_pool2d(input, output_size=(3, 3))
print(output.shape) # -> torch.Size([1, 3, 3, 3])
```
The derivatives assume there *is a batch dimension*, even if `batch_size = 1`.

Reply:
Nice test. I deleted output check and keep input check.